src/rrules/low_level_maths.jl

```diff
@@ -105,11 +99,93 @@ function rrule!!(::CoDual{typeof(log)}, b::CoDual{P}, x::CoDual{P}) where {P<:IE
     y = log(primal(b), primal(x))
     function log_adjoint(dy::P)
         log_b = log(primal(b))
-        return NoRData(), -dy * y / (log_b * primal(b)), dy / (primal(x) * log_b)
+        return NoRData(),
+        ifelse(iszero(primal(b)) || isnan(log_b), 0.0, -dy * y / (log_b * primal(b))),
+        ifelse(iszero(primal(x)) || isnan(log_b), 0.0, dy / (primal(x) * log_b))
     end
     return zero_fcodual(y), log_adjoint
 end
 
+@is_primitive MinimalCtx Tuple{typeof(log),P} where {P<:IEEEFloat}
+function frule!!(::Dual{typeof(log)}, x::Dual{P}) where {P<:IEEEFloat}
+    _x, dx = extract(x)
+    y = log(_x)
+    return Dual(y, dx / _x)
+end
+function rrule!!(::CoDual{typeof(log)}, x::CoDual{P}) where {P<:IEEEFloat}
+    y = log(primal(x))
+    function log_adjoint(dy::P)
+        return NoRData(), ifelse(iszero(primal(x)), 0.0, dy / primal(x))
+    end
+    return zero_fcodual(y), log_adjoint
+end
+
+@is_primitive MinimalCtx Tuple{typeof(sqrt),IEEEFloat}
+function frule!!(::Dual{typeof(sqrt)}, x::Dual{P}) where {P<:IEEEFloat}
+    _x, dx = extract(x)
+    return Dual(sqrt(_x), dx / (2 * sqrt(_x)))
```

Comment:
`sqrt(_x)` is called twice here -- it would be good to only compute it once, and use it twice.